text/3806-align-attr.md

```diff
@@ -282,12 +282,14 @@ This replaces `#[repr(align(…))]` on function items, from
 On `async fn`, the attribute controls the alignment of the code of the function
 that returns the `Future`.
 
-On function items in trait declarations, `#[align(…)]` specifies the minimum
-alignment that all implementations of the item must have. `impl` blocks
-containing the item must specify an `#[align(…)]` at least as high. The `dyn`
-implementation generated by the compiler must also provide this alignment. Any
-change to `#[align(…)]` on a function item in a trait declaration is therefore
-semver-breaking.
+On function items in trait declarations, `#[align(n)]` (for some alignment `n`)
+specifies the minimum alignment that all implementations of the item must have.
+The generated implementation for the trait’s `dyn` type will also provide this
+alignment. `impl` blocks do not have to repeat the `#[align(n)]` attribute, it
+is implicit. (This enables the trait definition to raise `n` without breaking
+implementations.) That being said, `impl` blocks are free to specify an
+`align(…)` higher than that provided by the trait. (If they specify a *lower*
+alignment, it will simply be ignored.)
```

Comment:
This at least resolves the semver hazard (which is an improvement). I'm still not sure what the use-case is for this such that we should allow it, since it mostly seems like it may inhibit optimizations by preventing unification of instances of the code, but in an unreliable way. But it does have consistency with `#[align]` specified twice at least.

Reply:
...Indeed it's actually not clear to me that we can prevent optimizations like identical code folding from collapsing code?

Reply:
Higher alignments fulfill lower alignments, so when merging together otherwise identical functions, the compiler can just keep the most-aligned one.

Reply:
That does not seem to be the semantic that LLVM actually implements: https://godbolt.org/z/7xK78v61f

Reply:
`@aligned_to_2p8()` is unused and therefore discarded. If you remove `@aligned_to_default()`, the output will end up entirely empty

Reply:
I have opened an issue about how we would prefer this semantic to be implemented: https://github.com/llvm/llvm-project/issues/145123

I believe there are mitigations we can pursue on our side, but it may come at notable cost.